node/core/approval-voting/src/lib.rs

```diff
@@ -683,9 +688,17 @@ where
 
 	let mut last_finalized_height: Option<BlockNumber> = None;
 
+	let mut startup_delay =
+		stream::once(futures_timer::Delay::new(STARTUP_WAIT_DURATION)).chain(stream::empty());
+
 	loop {
 		let mut overlayed_db = OverlayedBackend::new(&backend);
 		let actions = futures::select! {
+			_res = (&mut startup_delay).next() => {
+				let height = last_finalized_height.unwrap_or_default();
```

Comment:
I think finality notifications are skipped during major sync? It would be good if we waited not only for STARTUP_WAIT_DURATION, but also until `last_finalized_height` is `Some`.

Reply:
Should we implement some timeout such that if it doesn't turn to Some after N seconds we panic?

Reply:
No, we can't assume it takes N seconds to sync up to the tip (for major sync to stop).
We're issuing active-leaves on startup regardless https://github.com/paritytech/polkadot/blob/dee1484760aedfd699e764f2b7c7d85855f7b077/node/overseer/src/lib.rs#L751
but I believe not finality notifications. So if a validator is not synced, it will get its first finality notification once it's done syncing. @bkchr please correct me if I'm wrong

Reply:
Yes @ordian is right that we don't send finality nor import notifications on major sync.